File: services/trip-service/internal/application/dto/cost_splitting.go
package dto

import (
	"time"

	"github.com/google/uuid"

	"github.com/KovalenkoDima236961/Travel_Ai_App/internal/budget"
	"github.com/KovalenkoDima236961/Travel_Ai_App/internal/domain/aggregate"
	"github.com/KovalenkoDima236961/Travel_Ai_App/internal/domain/entity"
)

type CreateTripTravelerInput struct {
	Name         string
	Email        *string
	LinkedUserID *uuid.UUID
	Role         entity.TripTravelerRole
}

type UpdateTripTravelerInput struct {
	Name  *string
	Email *string
	Role  *entity.TripTravelerRole
}

type UpdateItemCostSplitInput struct {
	ExpectedItineraryRevision *int
	Split                     *aggregate.CostSplitRule
}

type UpdateAccommodationCostSplitInput struct {
	Split *aggregate.CostSplitRule
}

type CostSplittingSummary struct {
	TripID           uuid.UUID                `json:"tripId"`
	Currency         string                   `json:"currency"`
	GeneratedAt      time.Time                `json:"generatedAt"`
	Summary          CostSplittingTotals      `json:"summary"`
	Travelers        []TravelerCostAllocation `json:"travelers"`
	UnassignedCosts  []UnassignedCost         `json:"unassignedCosts"`
	ByCategory       []CostSplitCategoryTotal `json:"byCategory"`
	ByDay            []CostSplitDayTotal      `json:"byDay"`
	Warnings         []string                 `json:"warnings"`
	ExchangeRateInfo *budget.ExchangeRateInfo `json:"exchangeRateInfo,omitempty"`
}

type CostSplittingTotals struct {
	TravelerCount        int     `json:"travelerCount"`
	EstimatedTotal       float64 `json:"estimatedTotal"`
	AllocatedTotal       float64 `json:"allocatedTotal"`
	UnassignedTotal      float64 `json:"unassignedTotal"`
	MissingEstimateCount int     `json:"missingEstimateCount"`
	DefaultSplitCount    int     `json:"defaultSplitCount"`
	InvalidSplitCount    int     `json:"invalidSplitCount"`
	ConvertedItemCount   int     `json:"convertedItemCount"`
	UnconvertedItemCount int     `json:"unconvertedItemCount"`
}

type TravelerCostAllocation struct {
	TravelerID        uuid.UUID                `json:"travelerId"`
	Name              string                   `json:"name"`
	Email             *string                  `json:"email,omitempty"`
	LinkedUserID      *uuid.UUID               `json:"linkedUserId,omitempty"`
	Role              entity.TripTravelerRole  `json:"role"`
	AllocatedTotal    float64                  `json:"allocatedTotal"`
	PercentageOfTotal float64                  `json:"percentageOfTotal"`
	ByCategory        []CostSplitCategoryTotal `json:"byCategory"`
	ByDay             []CostSplitDayTotal      `json:"byDay"`
	Items             []TravelerAllocatedItem  `json:"items"`
}

type TravelerAllocatedItem struct {
	Type                 string  `json:"type"`
	DayNumber            *int    `json:"dayNumber,omitempty"`
	ItemIndex            *int    `json:"itemIndex,omitempty"`
	Name                 string  `json:"name"`
	Category             string  `json:"category"`
	AllocatedAmount      float64 `json:"allocatedAmount"`
	OriginalCostAmount   float64 `json:"originalCostAmount"`
	OriginalCostCurrency string  `json:"originalCostCurrency"`
	SplitType            string  `json:"splitType"`
	RuleSource           string  `json:"ruleSource"`
}

type UnassignedCost struct {
	Type      string  `json:"type"`
	DayNumber *int    `json:"dayNumber,omitempty"`
	ItemIndex *int    `json:"itemIndex,omitempty"`
	Name      string  `json:"name"`
	Amount    float64 `json:"amount"`
	Currency  string  `json:"currency"`
	Reason    string  `json:"reason"`
}

type CostSplitCategoryTotal struct {
	Category string  `json:"category"`
	Amount   float64 `json:"amount"`
}

type CostSplitDayTotal struct {
	DayNumber int     `json:"dayNumber"`
	Amount    float64 `json:"amount"`
}

File: apps/web/src/entities/cost-splitting/model/cost-splitting.ts
export type TripTravelerRole = "organizer" | "traveler";
export type TripTravelerStatus = "active" | "removed";

export type TripTraveler = {
  id: string;
  tripId: string;
  name: string;
  email?: string | null;
  linkedUserId?: string | null;
  role: TripTravelerRole;
  status: TripTravelerStatus;
  createdAt: string;
  updatedAt: string;
};

export type CreateTripTravelerInput = {
  name: string;
  email?: string | null;
  linkedUserId?: string | null;
  role: TripTravelerRole;
};

export type UpdateTripTravelerInput = {
  name?: string;
  email?: string | null;
  role?: TripTravelerRole;
};

export type CostSplitType = "all_equal" | "selected_equal" | "custom_percentages";

export type CostSplitRule = {
  type: CostSplitType;
  travelerIds?: string[];
  percentages?: Record<string, number>;
};

export type CostSplittingSummary = {
  tripId: string;
  currency: string;
  generatedAt: string;
  summary: CostSplittingTotals;
  travelers: TravelerCostAllocation[];
  unassignedCosts: UnassignedCost[];
  byCategory: CostSplitCategoryTotal[];
  byDay: CostSplitDayTotal[];
  warnings: string[];
  exchangeRateInfo?: CostSplitExchangeRateInfo | null;
};

export type CostSplitExchangeRateInfo = {
  provider?: string | null;
  asOf?: string | null;
  fallbackUsed?: boolean;
};

export type CostSplittingTotals = {
  travelerCount: number;
  estimatedTotal: number;
  allocatedTotal: number;
  unassignedTotal: number;
  missingEstimateCount: number;
  defaultSplitCount: number;
  invalidSplitCount: number;
  convertedItemCount: number;
  unconvertedItemCount: number;
};

export type TravelerCostAllocation = {
  travelerId: string;
  name: string;
  email?: string | null;
  linkedUserId?: string | null;
  role: TripTravelerRole;
  allocatedTotal: number;
  percentageOfTotal: number;
  byCategory: CostSplitCategoryTotal[];
  byDay: CostSplitDayTotal[];
  items: TravelerAllocatedItem[];
};

export type TravelerAllocatedItem = {
  type: "itinerary_item" | "accommodation" | string;
  dayNumber?: number | null;
  itemIndex?: number | null;
  name: string;
  category: string;
  allocatedAmount: number;
  originalCostAmount: number;
  originalCostCurrency: string;
  splitType: CostSplitType | string;
  ruleSource: "explicit" | "default" | string;
};

export type UnassignedCost = {
  type: "itinerary_item" | "accommodation" | string;
  dayNumber?: number | null;
  itemIndex?: number | null;
  name: string;
  amount: number;
  currency: string;
  reason: string;
};

export type CostSplitCategoryTotal = {
  category: string;
  amount: number;
};

export type CostSplitDayTotal = {
  dayNumber: number;
  amount: number;
};

export type TripTravelersResponse = {
  travelers: TripTraveler[];
};
